go/simple/v1/ampli/ampli.go

```diff
@@ -0,0 +1,385 @@
+			"and apiKey from either options.Client.APIKey or APIKey[options.Environment], " +
+			"or options.Client.Instance")
+	}
+
+	var configuration amplitude.Config
+	if !options.Client.Configuration.IsEmpty() {
+		configuration = options.Client.Configuration
+	} else {
+		configuration = DefaultConfiguration
+	}
+
+	if configuration.Plan == (amplitude.Plan{}) {
+		configuration.Plan = DefaultConfiguration.Plan
+	}
+
+	if options.Client.Instance != nil {
+		a.Client = options.Client.Instance
+	} else {
+		configuration.APIKey = apiKey
+		a.Client = amplitude.NewClient(configuration)
+	}
+}
+
+// InitializedAndEnabled checks if Ampli is initialized and enabled.
+func (a *Ampli) InitializedAndEnabled() bool {
+	return !a.Disabled
+}
+
+// Track tracks an StronglyTypedEvent.
+func (a *Ampli) Track(userID string, event StronglyTypedEvent, eventOptions amplitude.EventOptions) {
+	if !a.InitializedAndEnabled() {
+		return
+	}
+
+	if userID != "" {
+		eventOptions.UserID = userID
+	}
+
+	baseEvent := event.ToEvent()
+	baseEvent.EventOptions = eventOptions
+
+	a.Client.Track(baseEvent)
+}
+
+// Identify identifies a user and set user properties.
+func (a *Ampli) Identify(userID string, identify Identify, eventOptions amplitude.EventOptions) {
+	if !a.InitializedAndEnabled() {
+		return
+	}
+
+	if userID != "" {
+		eventOptions.UserID = userID
+	}
+
+	identifyEvent := amplitude.Event{
+		EventType: amplitude.IdentifyEventType,
+		EventProperties: map[string]interface{}{
+			"requiredNumber": identify.RequiredNumber,
+			"optionalArray":  identify.OptionalArray,
+		},
+		EventOptions: eventOptions,
+	}
+
+	a.Client.Track(identifyEvent)
+}
+
+// GroupIdentify identifies a group and set group properties.
+func (a *Ampli) GroupIdentify(groupType string, groupName string, group Group, eventOptions amplitude.EventOptions) {
+	if !a.InitializedAndEnabled() {
+		return
+	}
+
+	groupIdentifyEvent := amplitude.Event{
+		EventType: amplitude.GroupIdentifyEventType,
+		Groups:    map[string][]string{groupType: {groupName}},
+		EventProperties: map[string]interface{}{
+			"requiredBoolean": group.RequiredBoolean,
+			"optionalString":  group.OptionalString,
+		},
+		EventOptions: eventOptions,
+	}
+
+	a.Client.Track(groupIdentifyEvent)
+}
+
+// SetGroup sets group for the current user.
+func (a *Ampli) SetGroup(userID string, groupType string, groupName []string, eventOptions amplitude.EventOptions) {
+	if !a.InitializedAndEnabled() {
+		return
+	}
+
+	if eventOptions == (amplitude.EventOptions{}) {
+		return
+	}
+
+	if userID != "" {
+		eventOptions.UserID = userID
+	}
+
+	a.Client.SetGroup(groupType, groupName, eventOptions)
+}
+
+// Flush flushes events waiting in buffer.
+func (a *Ampli) Flush() {
+	if !a.InitializedAndEnabled() {
+		return
+	}
+
+	a.Client.Flush()
+}
+
+// Shutdown disables and shutdowns Ampli Client.
+func (a *Ampli) Shutdown() {
+	if !a.InitializedAndEnabled() {
+		return
+	}
+
+	a.Client.Shutdown()
+	a.Disabled = true
```

Comment:
The assignment isn't thread-safe.

Reply:
Could you give a hint why this is not thread-safe?

Reply:
Simplified example (`main.go`):
```
package main

import (
	"time"
)

type Ampli struct {
	Disabled bool
}

func (a *Ampli) Enabled() bool {
	return !a.Disabled
}

func (a *Ampli) Track() {
	if a.Disabled {
		return
	}
	// track
}

func (a *Ampli) Shutdown() {
	a.Disabled = false
}

func main() {
	ampli := &Ampli{}

	for i := 0; i < 10; i++ {
		go func() {
			for {
				ampli.Track()
			}
		}()
	}

	time.Sleep(time.Second)

	ampli.Shutdown()

	time.Sleep(time.Second)
}
```

Try run command:
`go run -race main.go`

Output:
```
$ go run -race main.go 
==================
WARNING: DATA RACE
Write at 0x00c00009e00f by main goroutine:
  main.(*Amplit).Shutdown()
      /home/falconandy/GolandProjects/awesomeProject1/main.go:23 +0xe4
  main.main()
      /home/falconandy/GolandProjects/awesomeProject1/main.go:39 +0xd9

Previous read at 0x00c00009e00f by goroutine 6:
  main.(*Amplit).Track()
      /home/falconandy/GolandProjects/awesomeProject1/main.go:16 +0x33
  main.main.func1()
      /home/falconandy/GolandProjects/awesomeProject1/main.go:32 +0x2e

Goroutine 6 (running) created at:
  main.main()
      /home/falconandy/GolandProjects/awesomeProject1/main.go:30 +0x4b
==================
Found 1 data race(s)
exit status 66
falconandy@MAIN-APC:~/GolandProjects/awesomeProject1$ go run -race main.go 
==================
WARNING: DATA RACE
Write at 0x00c00011600f by main goroutine:
  main.(*Ampli).Shutdown()
      /home/falconandy/GolandProjects/awesomeProject1/main.go:23 +0xe4
  main.main()
      /home/falconandy/GolandProjects/awesomeProject1/main.go:39 +0xd9

Previous read at 0x00c00011600f by goroutine 7:
  main.(*Ampli).Track()
      /home/falconandy/GolandProjects/awesomeProject1/main.go:16 +0x33
  main.main.func1()
      /home/falconandy/GolandProjects/awesomeProject1/main.go:32 +0x2e

Goroutine 7 (running) created at:
  main.main()
      /home/falconandy/GolandProjects/awesomeProject1/main.go:30 +0x4b
==================
Found 1 data race(s)
exit status 66
```

Reply:
Added a mutex for every occurrence of `Ampli.Disabled`. 

Tried the above example, it ended with `panic: send on closed channel`, which is caused by core sdk. Created a ticket for it: https://amplitude.atlassian.net/browse/AMP-59243